cmd/zfs/zfs_main.c

```diff
@@ -1691,6 +1696,54 @@ zfs_do_destroy(int argc, char **argv)
 		usage(B_FALSE);
 	}
 
+	/* check type before destroying if specified */
+	if (type_filter != NULL) {
+		zhp = zfs_open(g_zfs, argv[0], ZFS_TYPE_DATASET);
+		if (zhp != NULL) {
+			zfs_type_t type = zfs_get_type(zhp);
+			boolean_t type_matches = B_FALSE;
+
+			switch (type) {
+			case ZFS_TYPE_FILESYSTEM:
+				type_matches = (
+				    (strcmp(type_filter, "fs") == 0) ||
+				    (strcmp(type_filter, "filesystem") == 0));
+				break;
+			case ZFS_TYPE_VOLUME:
+				type_matches = (
+				    (strcmp(type_filter, "vol") == 0) ||
+				    (strcmp(type_filter, "volume") == 0));
+				break;
+			case ZFS_TYPE_SNAPSHOT:
+				type_matches = (
+				    (strcmp(type_filter, "snap") == 0) ||
+				    (strcmp(type_filter, "snapshot") == 0));
+				break;
+			case ZFS_TYPE_BOOKMARK:
+				type_matches =
+				    (strcmp(type_filter, "bookmark") == 0);
+				break;
+			default:
+				type_matches = B_FALSE;
+				break;
+			}
```

Comment:
We have another type parser in `zfs_do_list()`.  It would be good to unify them.